draft-ietf-oauth-rfc7523bis.xml

```diff
@@ -295,51 +293,83 @@
+    client-assertion-type%3Ajwt-bearer&
+  client_assertion=eyJ0eXAiOiJjbGllbnQtYXV0aGVudGljYXRpb24rand0IiwiYWx
+    nIjoiRVMyNTYiLCJraWQiOiIxNiJ9.
+  e2F1ZDpodHRwczovL2F1dGh6LmV4YW1wbGUubmV0LA[...omitted...].
+  cC4hiUPo[...omitted for brevity...]
+]]></artwork>
+      </figure>
 
+      <t>
+	In Section 3 of <xref target="RFC7523"/> (JWT Format and Processing Requirements),
 	Item 3, which describes the audience value,
 	is replaced by:
 	<list style="empty">
 	  <t>
 	    The JWT MUST contain an <spanx style="verb">aud</spanx>
-	    (audience) claim containing
-	    the issuer identifier <xref target="RFC8414"/>
-	    of the authorization server as its sole value.
-	    The authorization server MUST have an issuer identifier
-	    to be used with this specification.
-	    Unlike the <spanx style="verb">aud</spanx> value specified
-	    in <xref target="RFC7523"/>, there MUST be no value other than
-	    the issuer identifier of the intended authorization server
-	    used as the audience of the JWT;
-	    this includes that the token endpoint URL of the authorization server
-	    MUST NOT be used as an audience value.
-	    To simplify implementations,
-	    the <spanx style="verb">aud</spanx> claim value MUST
-	    be a JSON string, and not a single-valued JSON array.
-	    The authorization server MUST reject any JWT that does not
-	    contain its issuer identifier as its sole audience value.
+	    (audience) claim containing a value that
+	    identifies the authorization server as the intended audience.
+	    Two cases are differentiated:
+
+	    <list style="letters">
+	      <t>
+		For the authorization grant,
+		it is the responsibility of the client to use only audience values
+		that are specific to the authorization server being used.
+		This MAY be either
+		the issuer identifier of the authorization server or
+		the token endpoint URL of the authorization server.
+		The authorization server MUST reject any JWT that does not
+		contain its own identity as the intended audience,
+		and if the value of <spanx style="verb">aud</spanx> is an array,
+		all array values MUST identify the authorization server
+		as the intended audience.
+	      </t>
+	      <t>
+		For client authentication,
+		the <spanx style="verb">aud</spanx> (audience) claim value MUST use
+		the issuer identifier <xref target="RFC8414"/>
+		of the authorization server as its sole value.
+		The authorization server MUST have an issuer identifier
+		to be used with this specification.
+		Unlike the <spanx style="verb">aud</spanx> value specified
+		in <xref target="RFC7523"/>, there MUST be no value other than
+		the issuer identifier of the intended authorization server
+		used as the audience of the JWT;
+		this includes that the token endpoint URL of the authorization server
+		MUST NOT be used as an audience value.
+		To simplify implementations,
+		the <spanx style="verb">aud</spanx> claim value MUST
+		be a JSON string, and not a single-valued JSON array.
```

Comment:
> MUST be a JSON string, and not a single-valued JSON array.

I understand the intent, but I believe that the 3 extra lines of code at the authorization server are worth it and don't create any significant security concerns. Systems that issue tokens by default with an audience array would need to make massive changes to support this wording here.

For instance: Kubernetes Service Account Tokens do always have an audience array - never a single string. Forcing a string would mean that we'd break every OAuth client that uses it's service account as client auth.

Reply:
Thanks for your review, @arndt-s.  The decision to use a single representation for the single-valued audience reached consensus among those discussing the security issue and mitigations.

That said, it's good data that Kubernetes only sends array-valued audience values.  We should merge this before the OAuth session, but I encourage you to file on issue on this topic if you'd like the decision to be reconsidered.

Reply:
Also FWIW "aud mandated to be a string value only" was one of the areas of overreach mentioned in #5